src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -693,6 +693,21 @@ public async Task<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request,
                 // hook up a continuation that will log it.
                 if (sendRequestContentTask != null && !sendRequestContentTask.IsCompletedSuccessfully)
                 {
+                    // In case the connection is disposed, it's most probable that
+                    // expect100Continue timer expired and request content sending failed.
+                    // We're awaiting the task to propagate the exception in this case.
+                    if (Volatile.Read(ref _disposed) == 1)
+                    {
+                        if (async)
+                        {
+                            await sendRequestContentTask.ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            // No way around it here if we want to get the exception from the task.
```

Comment:
This might actually block, the `sendRequestContentTask` might still not be finished - the `Dispose` called from `SendRequestContentWithExpect100ContinueAsync` might have already propagated and manifested in an exception here while the method itself hasn't finished.

Reply:
The very similar race condition happens in HTTP/2 as well. The exception caused by:
https://github.com/dotnet/runtime/blob/bff243c86ca5e0a9467cbb4a8b9bad6aeb90513e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L236
might be propagated 'faster' that the task completion of `SendRequestBodyAsync` resulting in `HttpRequestException` instead of the original exception from `HttpContent` serialization.

Reply:
I feel inclined to forgo the blocking here and propagate the exception only if the task has already finished. Kind of best-effort, but not worth the complications. In the worst case scenario, the content exception will only get logged.
@scalablecory, @stephentoub any input on this?

Reply:
Is there any risk of user code throwing an exception here and us not bubbling it up?

Logging should only be relied on as an "also" where some other actionable exception/status is being communicated elsewhere.

Reply:
> is there any risk of user code throwing an exception here and us not bubbling it up? 

Yes, it happens in the tests on CI. It's what I'm describing above. We do unblock the main execution flow (`SendAsyncCore`), but the connection `Dispose` propagation might be faster than the completion of `sendRequestContentTask`.

As I said, very similar thing happens in HTTP/2. There wasn't and isn't any hang problem, but sometimes the exception from abort is faster than completion of request sending task and we bubble up the "wrong" exception. This already exists, I just caught it with my new test.

If we do want to bubble up the right exception (the one from request content sending), I'd need change the HTTP/2 flow as well.

Reply:
I also have an alternative solution, which tears down the connection after `sendRequestContentTask` finishes (inside the `expect100Continue` timer callback) but it requires using a closure for the callback (chicken-egg problem with `sendRequestContentTask` and the timer instance). I can open WIP PR with it if you want to see it?

Reply:
>> This might actually block, the sendRequestContentTask might still not be finished - the Dispose called from SendRequestContentWithExpect100ContinueAsync might have already propagated and manifested in an exception here while the method itself hasn't finished yet.

I might be missing something, but I don't see how `Dispose` could has already manifested itself in this case. Caller is waiting on this thread (in `await` or `GetResult()` and can only **here** observe an exception thrown from  `SendRequestContentWithExpect100ContinueAsync` -> `SendRequestContentAsync`. Or, are you talking about the case when `Dispose` has been called from some other place on a different thread?

Reply:
Yes, `SendRequestContentWithExpect100ContinueAsync` is triggered via `expect100Timer` from a different thread. It disposes the connection, unblocking the `SendAsyncCore` here, while the task itself hasn't finished yet.